drivers: ethernet: eth_stm32 add tx async mode

Freey0 opened the pull request:
Enable TX transmission asynchronous mode to enhance performance
    
    Tested on the nucleo_h743zi board using samples/net/zperf
    
    west build -b nucleo_h743zi/stm32h743xx samples/net/zperf/ --pristine
    -- -DCONFIG_NET_CONFIG_MY_IPV4_ADDR=\"192.168.1.199\"
    
    upload cmd:
    ```
    //linux
    iperf -s -l 1K -p 5001
    iperf -s -l 1K -p 5002 -u
    
    //shell
    zperf tcp upload 192.168.1.109 5001 10 1K
    zperf udp upload 192.168.1.109 5002 10 1K 300M
    ```
    
    download cmd:
    ```
    //shell
    zperf tcp download 5001 192.168.1.199
    zperf ucp download 5002 192.168.1.199
    
    //linux
    iperf -l 1K -c 192.168.1.199 -p 5001
    iperf -l 1K -c 192.168.1.199 -p 5002 -u
    ```

Freey0 commented:
After adding the configuration, the performance has approached the level of other RTOS.

```
# For TCP
CONFIG_PICOLIBC_USE_MODULE=y
CONFIG_ETH_STM32_HW_CHECKSUM=y

#For UDP
CONFIG_NET_ZPERF_MAX_PACKET_SIZE=1470
```
udp upload:      94.1Mbps/s
tcp upload:        92.9Mbps/s
udp download:  93.7Mbps/s
tcp download:   94.5Mbps/s

maass-hamburg commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -59,22 +59,6 @@ static const struct device *eth_stm32_phy_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0,
 #define ETH_STM32_AUTO_NEGOTIATION_ENABLE                                                          \
 	UTIL_NOT(DT_NODE_HAS_PROP(DT_INST_PHANDLE(0, phy_handle), fixed_link))
```
might be better to split the v1 and v2 driver completely.

Freey0 replied:
This information needs to be extracted from the device tree. Since DT_DRV_COMPAT does not differentiate between v1 and v2, there is no good way to incorporate this part into eth_stm32_hal_v1.c. 

Do you have any suggestions?

maass-hamburg replied:
both can have the same DT_DRV_COMPAT. V1 and V2 will not be used at the same time, that's depended on the soc.

Also please make it so, that the v2 driver will preserve the git history of the original driver, so that git blame works at least for v2. maybe rename it in one commit and then split v1 out into v1 driver file or something like that

Freey0 replied:
I used git mv, but still lost the history of the code in eth_stm32_hal.c and eth_stm32_hal_v1.c. 

```
git mv eth_stm32_hal.c eth_stm32_hal_v2.c
git commit -s

cp eth_stm32_hal_v2.c eth_stm32_hal_v1.c
cp eth_stm32_hal_v2.c eth_stm32_hal.c

clean code eth_stm32_hal_v2.c
clean code eth_stm32_hal_v1.c
clean code eth_stm32_hal.c

git add .
git commit -s
```
Are there any better ways to retain all the historical information?

Freey0 replied:
The special code v1 consists of only ~300 lines.

Freey0 replied:
It's not 300, it's 600.  Here is a new [PR 94142](https://github.com/zephyrproject-rtos/zephyr/pull/94142) that splits v1/v2 in this way. There is no common eth_stm32_hal.c. To be honest, there is quite a lot of repetitive code.

Freey0 commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -82,6 +82,12 @@ config ETH_STM32_HW_CHECKSUM
 	  performances.
 	  See reference manual for more information on this feature.
 
+config ETH_STM32_HAL_TX_ASYNC
```
From my own tests, I haven't found any problems so far. Or, to be more radical, would it be better to completely remove the synchronization path?

maass-hamburg replied:
If there are no drawbacks,yes

maass-hamburg commented on drivers/ethernet/eth_stm32_hal_v1.c:
```diff
@@ -0,0 +1,297 @@
+		k_yield();
+	}
+
+#if defined(CONFIG_PTP_CLOCK_STM32_HAL)
```
CONFIG_PTP_CLOCK_STM32_HAL depends on the v2 hal, v1 doesn't support ptp

Freey0 replied:
Will clean the ptp code in v1

Freey0 commented:
@erwango Could you help conduct comprehensive testing on the farm? Thank you very much.

Freey0 commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -81,6 +81,14 @@ config ETH_STM32_HW_CHECKSUM
 
+config ETH_STM32_HAL_TX_ASYNC
+	bool "Use TX async mode"
+	default y
```
How does one apply this to the entire sample? The overlay file is intended for a single board; how does one apply it to all STM32 boards?

erwango commented:
> @erwango Could you help conduct comprehensive testing on the farm? Thank you very much.

@juickar Can you take this in charge ? This being said, this should only be considered as a simple non regression.
Tests performed in CI are pretty basic, so manual extra testing would be required.

etienne-lms commented on drivers/ethernet/eth_stm32_hal_v2.c:
```diff
@@ -116,6 +116,122 @@ static inline uint16_t allocate_tx_buffer(void)
 
+#if defined(CONFIG_ETH_STM32_HAL_TX_ASYNC)
+/* allocate a tx context and mark it as used, the first tx buffer is also allocated */
+static inline struct eth_stm32_tx_context *allocate_tx_context_async(struct net_pkt *pkt)
```
For info, you can drop the `inline` keyword.

etienne-lms reviewed (commented):
Some implementation comments, aside performance impact.

Freey0 committed:
drivers: ethernet: eth_stm32 tx_in_sem limit set to 1

tx_int_sem serves solely as a signal indicator and does not require signal
counting. Within eth_stm32_tx, the semaphore is reset each time a packet
is transmitted. Upon receiving the transmission completion signal, the
function exits. Signal counting is unnecessary. Change to a binary
semaphore.

Signed-off-by: WenBin Zhang <freey7955@gmail.com>

Freey0 committed:
drivers: ethernet: eth_stm32 add tx async mode

Enable TX transmission asynchronous mode to enhance performance

Tested on the nucleo_h743zi board using samples/net/zperf

west build -b nucleo_h743zi/stm32h743xx samples/net/zperf/ --pristine
-- -DCONFIG_NET_CONFIG_MY_IPV4_ADDR=\"192.168.1.199\"

upload cmd:
```
//linux
iperf -s -l 1K -p 5001
iperf -s -l 1K -p 5002 -u

//shell
zperf tcp upload 192.168.1.109 5001 10 1K
zperf udp upload 192.168.1.109 5002 10 1K 300M
```

download cmd:
```
//shell
zperf tcp download 5001 192.168.1.199
zperf ucp download 5002 192.168.1.199

//linux
iperf -l 1K -c 192.168.1.199 -p 5001
iperf -l 1K -c 192.168.1.199 -p 5002 -u
```

before:
udp upload:   73.5Mbps/s
tcp upload:   71.3Mbps/s
udp download: 93.9Mbps/s
tcp download: 70.5Mbps/s

after:
udp upload:   92.2Mbps/s
tcp upload:   78.1Mbps/s
udp download: 93.7Mbps/s
tcp download: 75.2Mbps/s

Signed-off-by: WenBin Zhang <freey7955@gmail.com>

Freey0 committed:
samples: zperf: Add STM32 eth_tx_async mode test

enable STM32 etc_tx_async mode for testing in CI

Signed-off-by: WenBin Zhang <freey7955@gmail.com>

erwango commented on drivers/ethernet/Kconfig.stm32_hal:
```diff
@@ -81,6 +81,14 @@ config ETH_STM32_HW_CHECKSUM
 
+config ETH_STM32_HAL_TX_ASYNC
+	bool "Use TX async mode"
+	default n
```
Don"t explicit `default n`. This is the default behavior.

erwango commented on samples/net/zperf/sample.yaml:
```diff
@@ -17,6 +17,10 @@ tests:
     extra_configs:
       - CONFIG_NET_ZPERF_SERVER=n
+  sample.net.zperf.async_tx.stm32:
+    filter: dt_compat_enabled("st,stm32-ethernet")
```
To avoid errors on configuration that does not have the right dependencies, it should be:
```
    filter: dt_compat_enabled("st,stm32-ethernet") and CONFIG_ETH_STM32_HAL_API_V2
```
My bad for the initially wrong proposal.

erwango commented on samples/net/zperf/boards/nucleo_h743zi.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_ETH_STM32_HAL_TX_ASYNC=y
```
I can see you kept it in your lastest push. I won't block but would prefer to get rid of it.
